README.md
Advanced Wesnoth Wars
=======================

About
-----------

It's a mod  for "Battle for Wesnoth" open-source videogame.

* Author :  [Ruvaak](http://thomas.bondois.info)
* Mantainer : [Toranks](http://toranks.blogspot.com/)
* Version : see [lua/aww_status.lua:9](lua/aww_status.lua)
* Changelog : see [CHANGELOG.md](CHANGELOG.md)
* License : [GNU-GPL](LICENSE.md)
* Code Repository [on GitHub](https://github.com/Toranks/Advance_Wesnoth_Wars)
* Dedicated [forum thread](https://forums.wesnoth.org/viewtopic.php?f=15&t=49454)


DISCLAIMER FOR MODDERS
----------------------

You can reuse the code of this mod in your own creations, without asking my authorization, only if you share it with the same License, freely.
But I will really appreciate to know my code is used on your project, and that you credit my work somewhere in your code.


Description
-----------

Last Wesnoth version tested : 1.17.16

A collection of the most-requested options to make the mechanics more realistic and tactic, for people curious to experiment a new gameplay.
All theses features works for mainline Campaigns & Scenarios, Single or Multiplayer, for all sides including AI (except for special cases).

The features which keep the campaigns balanced are enabled by default. The others features are more aimed for PvP multiplayer, or to try greater challenges.


Features list
-----------

- No Random Combats: attacks never miss, terrain bonus is instead used as damage reducer. You will be able to change the base damage by a percentage to balance if necessary.

- Squad Mode (HP-related strikes) : Units do proportionally less attack strikes the less HP they have (or less damage for single-strike attacks). Swarm is a worse ratio (-1 strike if you miss just 1 HP), than native, but taken in account by AI. In both cases, minimum strikes will be 1, and will not apply for berserk.

- Level-Up : Increase Level Number : The unit level will continue growing with the AMLA level-ups. This affect to some abilities as leadership, upkeep cost, and experience the enemies get fighting or killing the unit. You can change the maximum level of leadership and inspire abilities with this feature active.

- Level-Up : Floating Notifications : Floating text on the unit currently leveling-up. Can be helpful to know which unit is concerned by an advancement.

- Level-Up : Relative Healing : When an unit advance a level, instead of a full healing and being cured, it keep the missing HP, but statuses are cured. The special case when post-advancement Max-HP is smaller than pre-advanced is also handled.

- *Updated* Level-Up After Max Level Advancement - Manual or Random Bonuses : Units already AMLA (purple XP bar) gain an extra random ability or increased stat each time they level-up again. To identify them, they will have a new bracelet icon. Possible rewards: Melee damage +3, Melee strikes +1, Ranged damage +2, Ranged strikes +1, Movement points +1, First strike, Poison, Leadership, Distract, Regenerates, Adrenaline, Backstab, Marskman, Loyal, Fearless, Slows or Camouflage.

- Level-Up After Max Level Advancement - Promoted Leaders : Standard units reaching their first AMLA level-up will be able to recruit the same units as the original leader, recall, and will receive a bronze crown icon (keeping the Loyal icon in case). You can set the minimum level to be promoted. Excluded for special Heroes.

- Epic Heroes : The Leaders and Heroes (gold & silver crowns) gain an increase of 20% of their attributes, and a small regeneration ability, making them harder to kill. They also get the distraction ability, that excludes Zone of Control around for allies, but not for themselves. Excluded for Promoted Leaders.

- *Updated* Learning from battlefield (Passive XP) : Extra XP for all player units each turn (except for the ones not recalled). You can choose Passive XP for low level units, for AI and for AI leaders separately.

- Learning from healing (Max XP/turn for Healing) : Each turn, Healers will earn 1 XP for each adjacent wounded or poisoned ally, but you can limit the max/turn.

- Berserk tweak - Fury : Replace 'berserk' by a new 'Fury' weapon special, way more interesting. A warcry related to the fury level with also be displayed. You can gain ability 'Bloodthirsty' by killing a certain number of enemies.

- *New* Berserk rounds: Change the combat rounds of standard berserk ability, making it less risky.

- *Updated* Ambushed tweak - Surprise Attacks : An ambush will trigger a quick combat in which the ambushed unit can't counter-attack. A new eye icon indicates if the surprise attack is prepared. To avoid free exploration, undo can be disabled for some actions all the time, or when there is at least one unit alive and in play with the ability to hide.

- Ninja Wars! (Stealthy units) : Most of units will get melee back-stabs, poison arrows, diversion, skirmisher, be faster and invisible/ambushers in villages, forests, deep water, or everywhere at night. For funny PvP with fog-of-war focused on ambushes / hide&seek.

- *New* Villages line of sight : Villages can have a line of sight for the owner side, with configurable distance.

- *New* Recall cost of units : Choose the cost to recall units with chosen level or lower. Cost of higher level units remains unchanged. You can select free cost, 10, recruit cost + 1, percentage of recruit cost, or no changes.

- *New* Back to reserve : Allows you to send normal units back to the recall list with right click on keeps. Also set the cost for performing this action.

- *New* Desertions : One unit desertion will occur per turn when a human player side has less gold than a value you can specify.

- *New* Help menu : Contextual help accesible with right-click anywhere, with detailed explanation of the features, and the status or configured value for each, with options to activate, deactivate, or change values.

- *New* Tools : Utility to setup an AI controller for sides, leaders, and individual units, change controller of any side to human or AI, activate Fast AI, change any side color, create labels with custom color, category and tooltip, recolor units randomly to add variety, change poison HP damage, HP recovered by resting, default recall cost, base gold income, village income, village support, experience multiplier and experience gained fighting or killing units.

- *New* Import and export units anywhere: Contextual menu accesible with right-click, with capability of import and export units from and to any campaign or scenario.

- Guaranteed compatibility with: All mainline campaigns, To Lands Unknown, A New Order, Legends of the Invincibles, Non-Fatal Wounds Extended, and most of the existing campaigns, eras and modifications.

- Also includes a PvPvE scenario : '4p - Ruvaak Mirage Atoll' optimized to be combined with all features.

Feel free to contribute !
If you experiment a bug, please report it to me on the forum, then try to disable the feature (see next section).


How To Enable/Disable Mod Options during a Scenario/Campaign (for developers)
-----------------------------------------------

Note: This guide is kept for developers, but for normal users it is no longer necessary due to the new Help Menu, which covers almost all these functions.

When you disable an option, *MOST* of the changes made on units by the features are removed. But it's experimental.

1) During a game, type :
```
:debug
``` 
(except if you launched the game with --debug argument)

2) Then open Lua Console(default : `, but you can see it in Menu > Settings > Shortcuts > Display lua console)

3) Copy/paste/execute one of the following lines related to mod features. 

    - Change the number in parentheses : 0 (zero) means disabled, 1 enabled.
	
    - When specified, you can replace it by another number. (like in feature 02, 04, 05, 08...)
	
    - Most of the changes will be operational immediately, or next turn, or next scenario. But units already modified will keep theirs changes.
	
The commands to edit each feature and remove/re-enable changes on units :
```lua
aww_status.update_feature_01(0) -- NoRandomCombats
aww_status.update_feature_02(0) -- Squad Mode 1 = custom, 2 = swarm
aww_status.update_feature_03(0) -- L-Up AMLA Promoted Leader
aww_status.update_feature_04(0) -- passive xp, 0 to 6
aww_status.update_feature_05(0) -- healing xp 0 to 6
aww_status.update_feature_06(0) -- L-Up  relative healing
aww_status.update_feature_07(0) -- verbose debug messages
aww_status.update_feature_08(0) -- NoRandomCombats Damages Adjustment, to -20 to 40
aww_status.update_feature_09(0) -- AMLA Random Bonuses
aww_status.update_feature_10(0) -- Epic Leaders
aww_status.update_feature_11(0) -- Ninja Wars
aww_status.update_feature_12(0) -- Berserk tweak
aww_status.update_feature_13(0) -- Ambush tweak
aww_status.update_feature_14(0) -- L-Up Notif
aww_status.update_feature_15(0) -- L-Up AMLA number
```
If you want to display all currently enabled features (and the associated number), copy/paste/execute this line the lua console :
```lua
aww_status.message_info_all()
```
To display only activates features on a compact format, use :
```lua
aww_status.message_info()
```

You can set fury kills needed to gain bloodthirsty on any moment writing this with debug mode activated:
```lua
:set_var aww_12_fury_kills=X
```
To change Passive XP for AI:
```lua
:set_var aww_04_passive_xp_ai=X
```
A similar method can be used to edit any of the variables used by this addon.


How To Migrate a Savegame from a previous version of the mod
-------------------------------

It will do it automatically on next scenario. If you want to try to migrate the current scenario, copy/paste in lua console the following line :
```lua
wesnoth.fire_event('aww_event_reload_lua')
```
If you see some bugs on a feature,  (see previous section) disable it on re-enable it (replace XX by feature number):
```lua
aww_status.update_feature_XX(0)
aww_status.update_feature_XX(1)
```
See previous section to know how to open console or when the changes will applies.


Developer Mode
--------------

This is useful only for development/test purpose on the addon lua scripts. 

- Do not do that play a big campaign, as it can create issue with your savegame when you update the addon.
- Effect : the LUA scripts will be loaded externally (means you can edit them, and the effect will apply when you reload a savegame, no need to re-create a game or a scenario like for WML scripts).
- To enable it, just place a empty file named :
```
aww.dev
```
in the addon base directory (where the _main.cfg is).

- To disable it, just remove the file, and restart the game.


Thanks
-----------
Some of theses features are inspired from existing code, saving me lot of times.

A big thanks particularly to Eagle 11, Ravana, Bob_The_Mighty, Dugi, for their creation.

And a huge thanks for all developers and Wesnoth WML wiki contributors !

(Toranks): I say the same and add thanks to CelticMinstrel, Octalot and Pentarctagon without whom the code would not have improved as much as it has in recent versions.


CHANGELOG.md
# Changelog

For more clarity about compatibility, first 2 version numbers will follow base game version.

So 1.14.X.Y means game was checked compatible with Wesnoth 1.14. X is incremented for new features; Y is increment for fixes/optimisations.

## TO-DO
- Split the addon on different resources, to load only the features that interest the player. Probably not neccesary, so far I have not received complaints of poor performance or problems for unifying all in one addon.


## [1.17.18.3] - 2023-7-XX

### Changed


### Fixed
- Fixed amla bonus malfunction because a typo.


## [1.17.18.2] - 2023-7-13

### Changed
- Half cost of recall reflects exact value on Help Menu.
- Added Solid Phoenix to Epic Heroes.
- Removed regeneration on Beelzebub with Epic trait.

### Fixed
- [effect] apply_to=level compatible with LOTI,AtS,IftU and WoL.
- AI Controller: frozen and moving_to_target statuses changed to unit variables, to avoid deletion on advancements or others situations.
- Fixed amla bonus malfunction because a typo.


## [1.17.18.1] - 2023-6-30

### Changed
- Standardized print messages colors. Green=Activate function, Red=Deactivate function, Cyan=Change value or info.
- Added print messages to silent features or options.

### Fixed
- Some strings now are translatable. Spanish translation included.
- Desertions: Better detection of loyals.


## [1.17.18] - 2023-6-19

### Changed
- Better detection of ranged attacks on Surprise Attack.

### Fixed
- Removed menu items on leader AI controller between scenarios.


## [1.17.17.2] - 2023-6-6 on 1.16 as 1.17.17.1

### Changed
- Ambush icon now it is not reset in each scenario, but is preserved from one scenario to another. Only the counter variables are reset.

### Fixed
- Prevent gaining healing experience on side 1 first turn.


## [1.17.17.1] - 2023-6-2 on 1.17 only

### Changed
- Removed unbalanced +2 HP per level on Epic trait.

### Fixed
- Corrections on help menu.
- Better leader detection for Leader AI.
- Disallow undo on recall units with ambush icon only on LOTI, to avoid sync issues with undo+redo.


## [1.17.17] - 2023-5-21

### Added
- New Option: Ambush Tweak - Disallow Undo. Allows to disable undo of some actions always or when there is present a unit with hides.
- New Option: Learning XP - Excluded sides. Allows to exclude any side from receiving any form of learning XP.

### Changed
- Increased max experience per turn for human and AI to 999
- Added the merchant to units that do not receive Passive Experience.

### Fixed
- Various corrections on help menu.
- New [print2] tag to some sections to prevent it from appearing hidden behind the message. Only compatible with 1.17.3 and later.


## [1.17.16] - 2023-5-5

### Added
- Arrow icon for all options without icon, now is more readable

### Changed
- New and better scaled icons on contextual menu.
- Unit Controller can have the same side on controller and controlled, as the other AI Controllers, just for special cases.
- Double max XP increase on loyal Manual Bonus, and melee and ranged strike +1 on units with 1 strike on the base attack.
- Reduced loyal probability with Random Bonuses.
- Deactivated stop order for leaders when a move order is active.

### Fixed
- Now "Send to reserve" can be activated (and deactivated) on any moment instead of the next scenario.
- Unit controller menu properly cleaned between scenarios.
- Fixed icon collision with Non Fatal Wounds (harmless).
- Fixed #undef warnings of AI Controller.
- Fixed status of Leader AI (keep or engage) between scenarios.


## [1.17.15] - 2023-4-26

### Added
- New Tool: Experience modifier, similar to experience_modifier key of [scenario], but can be changed at any moment.

### Changed
- Disable "move" menu on frozen leaders, forcing to clear the frozen state manually (same as Unit AI Controller). This also can reduce the number of menu options.
- Regeneration and distraction for Epic Heroes defaults to no.
- Changed [effect] apply_to=level to native support on 1.17.15 and higher.

### Fixed
- Delete new menu items (engage-disengage) on Leader AI Controller between scenarios.
- Fixes on AI Controller menu text.


## [1.17.14.4] - 2023-4-9

### Added
- New Tool: Unit AI Controller, to control individual units.
- Option to move Setup AI Controller to the Main menu or Tools menu, for comfort.
- New Event: LOTI P1C2S8, choose to give Epic trait to all Heroes or only a selection of the best.

### Changed
- Reorganized AI Controller menu, with better explanation of all features, step by step.

### Fixed
- Non Random Combat, Fixed wrong display of strikes on estimations when a [attacks] special is applied to opponent and not to himself. This doesn't affect to real attacks, only to estimations display.
- Fixed wrong portrait of leader on Side AI Controller on certain situations.
- More reliable detection of leader on Leader AI Controller, taking into account if a side does not have any leader or unit capable of recruiting, displaying an error message.
- Deleted retreat_factor AI aspect too when side behavior is cleared.


## [1.17.14.3] - 2023-4-3

### Added
- New Feature: Reload items on Legends of the Invincibles. (Only available on this campaign)
- New Tool option: Change controller to AI for only one turn.

### Fixed
- French translation rebuilt and ordered, 28%
- Minor fixes with texts and variables.


## [1.17.14.2] - 2023-4-2

### Fixed
- Better warning about Side Controller and how to regain control of sides if all sides belong to AI.
- Fixed Ambush Tweak on LOTI, that on some situations still need to re-equip items or attack at least once to rebuild the item effects. Better and cleaner code to that.


## [1.17.14.1] - 2023-4-1

### Added
- New Tool: Random Color. All units will receive a random tint of color, completely configurable in ranges of red, green, blue and opacity.
- New Leader AI tool options: Engage combat or take refuge on a keep.

### Fixed

lua/aww_status.lua
--<<
-- Author: Ruvaak
-- Original Creation

local _ = wesnoth.textdomain 'aww'

aww_status = {
	title = _"Advanced Wesnoth Wars",
	version = "1.17.18.2", -- same as in _server.pbl
	FEATURE_IDS = {
		[1]  = "aww_01_enable_randomless_combats",
		[2]  = "aww_02_squad_mode",
		[3]  = "aww_03_enable_promoted_leader",
		[4]  = "aww_04_passive_xp",
		[5]  = "aww_05_healing_xp",
		[6]  = "aww_06_enable_levelup_relative_healing",
		[7]  = "aww_07_enable_verbose",
		[8]  = "aww_08_randomless_damage_adjustment",
		[9]  = "aww_09_enable_amla_bonus",
		[10] = "aww_10_enable_gifted_leaders",
		[11] = "aww_11_enable_ninja",
		[12] = "aww_12_enable_berserk_fix",
		[13] = "aww_13_enable_ambushed_fix",
		[14] = "aww_14_enable_levelup_notif",
		[15] = "aww_15_enable_levelup_amla_inc",
		[16] = "aww_04_passive_xp_ai",
	},
}


function aww_status.run()
	aww_status.versioning()
	aww_status.init()
	-- aww_status.message_info()
end


function aww_status.init()
	aww_status.feature_01 = aww_status.get_feature_value(1, false)
	aww_status.feature_02 = aww_status.get_feature_value(2, 0)
	aww_status.feature_03 = aww_status.get_feature_value(3, false)
	aww_status.feature_04 = aww_status.get_feature_value(4, 0)
	aww_status.feature_05 = aww_status.get_feature_value(5, 0)
	aww_status.feature_06 = aww_status.get_feature_value(6, false)
	aww_status.feature_07 = aww_status.get_feature_value(7, false)
	aww_status.feature_08 = aww_status.get_feature_value(8, 0)
	aww_status.feature_09 = aww_status.get_feature_value(9, false)
	aww_status.feature_10 = aww_status.get_feature_value(10, false)
	aww_status.feature_11 = aww_status.get_feature_value(11, false)
	aww_status.feature_12 = aww_status.get_feature_value(12, false)
	aww_status.feature_13 = aww_status.get_feature_value(13, false)
	aww_status.feature_14 = aww_status.get_feature_value(14, false)
	aww_status.feature_15 = aww_status.get_feature_value(15, false)
	aww_status.feature_16 = aww_status.get_feature_value(16, 0)
end


function aww_status.versioning()
	local prev_version = wml.variables["aww_version"]
	if prev_version ~= nil and prev_version ~= aww_status.version then
		aww_status.migrate(prev_version)
	end
	wml.variables["aww_version"] = aww_status.version
end


-- migrate old versions name to keep options through versions of AWW on same campaign :
function aww_status.migrate(from_version)
	if aww_status.get_feature_value(1, nil) == nil and wml.variables["aww_enable_randomless_combats"] ~= nil then
		aww_status.update_feature_01(wml.variables["aww_enable_randomless_combats"])
	end
	if aww_status.get_feature_value(2, nil) == nil and wml.variables["aww_enable_health_based_combats"] ~= nil then
		local was_enabled_02 = wml.variables["aww_enable_health_based_combats"]
		if was_enabled_02 == true then
			aww_status.update_feature_02(1)
		else
			aww_status.update_feature_02(0)
		end
	end
	if aww_status.get_feature_value(3, nil) == nil and wml.variables["aww_enable_leader_promotions"] ~= nil then
		aww_status.update_feature_03(wml.variables["aww_enable_leader_promotions"])
	end
	if aww_status.get_feature_value(4, nil) == nil and wml.variables["aww_passive_xp"] ~= nil then
		aww_status.update_feature_04(wml.variables["aww_passive_xp"]*1)
	end
	if aww_status.get_feature_value(5, nil) == nil and wml.variables["aww_healing_xp"] ~= nil  then
		aww_status.update_feature_05(wml.variables["aww_healing_xp"]*1)
	end
	if aww_status.get_feature_value(6, nil) == nil and wml.variables["aww_enable_advance_healing_reduced"] ~= nil then
		aww_status.update_feature_06(wml.variables["aww_enable_advance_healing_reduced"])
	end
	if aww_status.get_feature_value(7, nil) == nil and wml.variables["aww_enable_debug"] ~= nil  then
		aww_status.update_feature_07(wml.variables["aww_enable_debug"])
	end
	if aww_status.get_feature_value(8, nil) == nil and wml.variables["aww_randomless_terrain_adjustment"] ~= nil then
		aww_status.update_feature_08(wml.variables["aww_randomless_terrain_adjustment"]*1)
	end
	if aww_status.get_feature_value(9, nil) == nil and wml.variables["aww_enable_amla_extra_bonus"] ~= nil  then
		aww_status.update_feature_09(wml.variables["aww_enable_amla_extra_bonus"])
	end
	if aww_status.get_feature_value(10, nil)== nil and wml.variables["aww_enable_gifted_leaders"] ~= nil  then
		aww_status.update_feature_10(wml.variables["aww_enable_gifted_leaders"])
	end
	if aww_status.get_feature_value(11, nil) == nil and wml.variables["aww_enable_stealth_mode"] ~= nil  then
		aww_status.update_feature_11(wml.variables["aww_enable_stealth_mode"])
	end

	wesnoth.interface.add_chat_message( _ "AWW notice", string.format( _ " Migrating from version %s to %s", wml.variables["aww_version"], aww_status.version))
	wml.variables["aww_migrated_version"] = from_version
end


function aww_status.get_info(filter_id)
	-- filter_id is an optionnal parameter. if not present/nil, will display all enabled features in the same order as in options menu
	local msg = ""
	if (filter_id == nil or filter_id == 1) and aww_status.feature_01 then
		msg = "01. ".._"No Random Combats" .. " | "
	end
	if aww_status.feature_01 and (filter_id == nil or filter_id == 1) and aww_status.feature_08 ~= 0 then
		msg = msg .. "08. ".._"Increased Damage"
				.. string.format(" [%s%%]",aww_status.feature_08)
				.. " | "
	end
	if (filter_id == nil or filter_id == 2) and aww_status.feature_02 and aww_status.feature_02 ~= 0 then
		local label = ""
		if aww_status.feature_02 == 1 then
			label = _"Custom"
		elseif aww_status.feature_02 == 2 then
			label = _"Swarm"
		end
		msg = msg .. "02. ".._"Squad Mode"
				.. string.format(" [%s:", aww_status.feature_02)
				..  label
				.. "] | "
	end
	if (filter_id == nil or filter_id == 6) and aww_status.feature_06 then
		msg = msg .. "06. ".. _"L-Up ".. _"Relative Heal" .. " | "
	end
	if (filter_id == nil or filter_id == 14) and aww_status.feature_14 then
		msg = msg .. "14. ".. _"L-Up ".._"Notif" .. " | "
	end
	if (filter_id == nil or filter_id == 15) and aww_status.feature_15 then
		msg = msg .. "15. ".. _"L-Up ".._"AMLA".." ".. _"Increase Level Number" .. " | "
	end
	if (filter_id == nil or filter_id == 9) and aww_status.feature_09 then
		msg = msg .. "09. ".. _"L-Up ".._"AMLA".." ".._"Bonuses" .. " | "
	end
	if (filter_id == nil or filter_id == 3) and aww_status.feature_03 then
		msg = msg .. "03. ".. _"L-Up ".._"AMLA".." ".._"Promoted Leaders" .. " | "
	end
	if (filter_id == nil or filter_id == 10) and aww_status.feature_10 then
		msg = msg .. "10. ".. _"Epic Heroes" .. " | "
	end
	if (filter_id == nil or filter_id == 4) and aww_status.feature_04 and aww_status.feature_04 ~= 0 then
		msg = msg .. "04. ".. _"Passive XP"
		 .. string.format(" [%s", aww_status.feature_04 )
		.. _"/turn]"
		 .. " |"
		 .. string.format(" [%s", aww_status.feature_16 )
		.. _"/turn AI]"
		 .. " | "
	end
	if (filter_id == nil or filter_id == 5) and aww_status.feature_05 and aww_status.feature_05 ~= 0
	then msg = msg .. "05. ".._"Healing XP"
			.. _" [Max "
			.. string.format("%s", aww_status.feature_05)
			.. _"/turn]"
			.. " | "
	end
	if (filter_id == nil or filter_id == 12) and aww_status.feature_12 then
		msg = msg .. "12. ".._"Berserk tweak" .. " | "
	end
	if (filter_id == nil or filter_id == 13) and aww_status.feature_13 then
		msg = msg .. "13. ".._"Ambush tweak" .. " | "
	end
	if (filter_id == nil or filter_id == 11) and aww_status.feature_11 then
		msg = msg .. "11. ".._"NINJA WARS!" .. " | "
	end
	if (filter_id == nil or filter_id == 7) and aww_status.feature_07 then
		msg = msg .. "07. ".._"Verbose"
	end

	if msg ~= "" then
		msg = _"Enabled :" .. " " .. msg
	elseif filter_id then
		msg = aww_status.get_feature_info(filter_id)
	else
		msg = _"All mod options are disabled."
	end
	return msg
end


function aww_status.get_feature_info(id)
	local msg = string.format(" #%s. %s = [%s]", id, aww_status.feature_code(id), aww_status.get_feature_value(id, '?'))
    msg = string.gsub(msg, "true", "ON")
    msg = string.gsub(msg, "false", "OFF")
    msg = string.gsub(msg, "aww_", "")
    msg = string.gsub(msg, "_", " ")
	return _"Feature" .. msg
end


function aww_status.message_info(id)
	wesnoth.interface.add_chat_message(aww_status.title .. " v".. aww_status.version, aww_status.get_info(id))
end


function aww_status.message_info_all()
	local msg = ""
	for id = 1, #aww_status.FEATURE_IDS do
		msg = msg .. aww_status.get_feature_info(id) .. " | "
	end
	wesnoth.interface.add_chat_message(aww_status.title .. " v".. aww_status.version, msg)
end


function aww_status.feature_code(id)
	if aww_status.FEATURE_IDS[id] ~= nil then
		return aww_status.FEATURE_IDS[id]
	end
	return "undefined"
end

function aww_status.get_feature_value(id, nil_value)
	local value = wml.variables[aww_status.feature_code(id)]
	if value == nil then
		 return nil_value
	end
	return value
end

function aww_status.set_feature_value(id, value)
	wml.variables[aww_status.feature_code(id)] = value
end

function aww_status.update_feature_01(value)
	local id = 1
	value = aww_status.to_bool(value)
	aww_status.feature_01 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_01_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		wesnoth.fire_event("aww_event_01_02_reset")
		if value then
			wesnoth.fire_event('aww_event_reload_duel')
		end
	end
	return value
end


function aww_status.update_feature_02(value)
	local id = 2
	value = value *1
	aww_status.feature_02 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_02_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		wesnoth.fire_event("aww_event_01_02_reset")
		if old_value == 0 then
			wesnoth.fire_event('aww_event_reload_duel')
		end
		if old_value == 2 then
			wesnoth.fire_event("aww_event_02_swarm_disable")
		end
		if value == 2 then
			wesnoth.fire_event("aww_event_02_swarm_enable")
		end
	end
	return value
end


function aww_status.update_feature_03(value)
	local id = 3
	value = aww_status.to_bool(value)
	aww_status.feature_03 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_03_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if value then
			wesnoth.fire_event("aww_event_03_enable")
		else
			wesnoth.fire_event("aww_event_03_disable")
		end
		if value == false then
		end
	end
	return value
end

function aww_status.update_feature_04(value)
	local id = 4
	value = value *1
	aww_status.feature_04 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_04_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
	end
	return value
end

function aww_status.update_feature_05(value)
	local id = 5
	value = value *1
	aww_status.feature_05 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value  then
		aww_status.feature_05_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
	end
	return value
end

function aww_status.update_feature_06(value)
	local id = 6
	value = aww_status.to_bool(value)
	aww_status.feature_06 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_06_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
	end
	return value
end

function aww_status.update_feature_07(value)
	local id = 7
	value = aww_status.to_bool(value)
	aww_status.feature_07 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_07_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if (aww_status.feature_01 or aww_status.feature_02 > 0) then
			wesnoth.fire_event('aww_event_reload_duel')
		end
	end
	return value
end

function aww_status.update_feature_08(value)
	local id = 8
	value = value *1
	aww_status.feature_08 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_08_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if (aww_status.feature_01) then
			wesnoth.fire_event('aww_event_reload_duel')
			wesnoth.fire_event("aww_event_01_02_reset")
		end
	end
	return value
end

function aww_status.update_feature_09(value)
	local id = 9
	value = aww_status.to_bool(value)
	aww_status.feature_09 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_09_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if value == false then
			wesnoth.fire_event("aww_event_09_disable")
		end
	end
	return value
end

function aww_status.update_feature_10(value)
	local id = 10
	value = aww_status.to_bool(value)
	aww_status.feature_10 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_10_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if value then
			wesnoth.fire_event("aww_event_10_enable")
		else
			wesnoth.fire_event("aww_event_10_disable")
		end
	end
	return value
end

function aww_status.update_feature_11(value)
	local id = 11
	value = aww_status.to_bool(value)
	aww_status.feature_11 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_11_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if value then
			wesnoth.fire_event("aww_event_11_enable")
		else
			wesnoth.fire_event("aww_event_11_disable")
		end
	end
	return value
end


function aww_status.update_feature_12(value)
	local id = 12
	value = aww_status.to_bool(value)
	aww_status.feature_12 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_12_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		if value then
			wesnoth.fire_event("aww_event_12_enable")
		else
			wesnoth.fire_event("aww_event_12_disable")
		end
	end
	return value
end


function aww_status.update_feature_13(value)
	local id = 13
	value = aww_status.to_bool(value)
	aww_status.feature_13 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_13_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
		wesnoth.fire_event("aww_event_13_reset")
	end
	return value
end

function aww_status.update_feature_14(value)
	local id = 14
	value = aww_status.to_bool(value)
	aww_status.feature_14 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_14_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
	end
	return value
end

function aww_status.update_feature_15(value)
	local id = 15
	value = aww_status.to_bool(value)
	aww_status.feature_15 = value
	local old_value = aww_status.get_feature_value(id, nil)
	if value ~= old_value then
		aww_status.feature_13_was = old_value
		aww_status.set_feature_value(id, value)
		aww_status.message_info(id)
	end
	return value
end


function aww_status.to_bool(v)
	if (type(v) =="boolean") then
		return v
	end
    return v and (
		(type(v) =="number" and (v > 0))
		or (type(v)=="string" and v ~= "false" and v ~= "no" and v ~= "0" and v ~= "")
	)
end


-- DEPRECATED alias for run(), for retrocompatibility in dev mode :
function aww_status.welcome()
	aww_status.run()
end


-- CUSTOM WML TAGS :

-- used in mods/10_gifted_leader_hero.cfg, mods/11_ninja.cfg
function wesnoth.wml_actions.aww_remove_trait(cfg)
	local trait_id = cfg.trait_id
	for _,unit in ipairs(wesnoth.units.find(cfg)) do
		wesnoth.units.remove_modifications(unit, {id = trait_id}, "trait")
	end
end

function wesnoth.wml_actions.aww_remove_object(cfg)
	local object_id_list = stringx.split(cfg.object_id)
	for _,unit in ipairs(wesnoth.units.find(cfg))
	do
		for _,object_id in ipairs(object_id_list)
		do
			wesnoth.units.remove_modifications(unit, {id = object_id})
		end
	end
end

aww_status.run() -- load aww_status.up.lua instead
>>
